### PR TITLE
merge types with same structure in remote schema (fix #1112)

### LINKED ISSUE #1112: merge remote types if their name and structure are same
In remote schemas, right now we don't merge types if they have the same name. If the types have same name and exact same structure we should merge them.

### PR DESCRIPTION
**Pull request by ecthiender**
### Description
If the remote schema has types with same structure and same name, they are merged now.

What component does this PR affect? 

- [x] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System

Requires changes from other components? If yes, please mark the components:

- [ ] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System

### Related Issue

#1112 

### Solution and Design

We have a `TyEq` typeclass. Declared instances of `TyEq` on various `TypeInfo` types, which compares the structure except `name` and `TypeLoc`.
`ObjTyInfo` is compared by name and its fields.

### Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update
- [ ] Community content

### Checklist:

- [x] I have read the **[contributing guide](https://github.com/hasura/graphql-engine/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
- [x] This change requires a change in the documentation. 
- [x] I have updated the documentation accordingly.
- [x] I have added required tests.

### CODE REVIEW

File: server/src-lib/Hasura/GraphQL/Validate/Types.hs [hunk 1]
```diff
@@ -38,6 +39,8 @@ module Hasura.GraphQL.Validate.Types
   , fromSchemaDocQ
   , TypeMap
   , TypeLoc (..)
+  --, TyEq (..)
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Validate/Types.hs]: remove the comment

File: server/src-lib/Hasura/GraphQL/Validate/Types.hs [hunk 2]
```diff
@@ -63,6 +66,16 @@ import           Hasura.RQL.Types.RemoteSchema
...
 import           Hasura.SQL.Value
 
+
+-- | Typeclass for equating relevant properties of various GraphQL types
+-- | defined below
+class EquatableGType a where
+  type EqProps a
+  getEqProps :: a -> EqProps a
+
+typeEq :: (Show (EqProps a), EquatableGType a, Eq (EqProps a)) => a -> a -> Bool
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Validate/Types.hs]: Why the `Show` constraint?

Comment (ecthiender) [reply server/src-lib/Hasura/GraphQL/Validate/Types.hs]: I was debugging. Will remove it.

File: server/src-lib/Hasura/GraphQL/Validate/Types.hs [hunk 3]
```diff
@@ -172,6 +197,10 @@ data InpObjTyInfo
   , _iotiLoc    :: !TypeLoc
   } deriving (Show, Eq, TH.Lift)
 
+instance EquatableGType InpObjTyInfo where
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Validate/Types.hs]: So we are comparing the descriptions of the fields in `InpObjTyInfo`?

**Comment by 0x777** [comment]
@ecthiender This also closes https://github.com/hasura/graphql-engine/issues/1135?

**Pull request merged**